pkg/vsphere/tasks/waiter.go

```diff
@@ -27,12 +28,47 @@ import (
 	"github.com/vmware/govmomi/vim25/progress"
 	"github.com/vmware/govmomi/vim25/soap"
 	"github.com/vmware/govmomi/vim25/types"
+	"github.com/vmware/vic/pkg/trace"
 )
 
 const (
 	maxBackoffFactor = int64(16)
+
+	VMContextObjectKey = "VMContextObject"
+)
+
+var (
+	m             sync.RWMutex
+	errorHandlers []ErrorHandler
 )
 
+type ErrorHandler interface {
+	// Handle returns true if the error is expected. Otherwise, return false. If any error happens during error handling, return (true, error).
+	Handle(ctx context.Context, err error) (bool, error)
+}
+
+type TaskInProgressHandler struct{}
+
+func (h *TaskInProgressHandler) Handle(ctx context.Context, err error) (bool, error) {
+	if isTaskInProgress(err) {
+		// TaskInProgress error, no need to fail here
+		log.Debugf("TaskInProgress error, continue")
+		return true, nil
+	}
+	return false, nil
+}
+
+func init() {
+	RegisterErrorHandler(&TaskInProgressHandler{})
+}
+
+func RegisterErrorHandler(handler ErrorHandler) {
```

Comment:
With a global list of error handlers, callers to WaitForResult may get unexpected results, since multiple handlers may be called, and handling depends on the order they were added. Why don't we just have another argument to WaitForResult for the handler?